pkg/sql/lease.go

```diff
@@ -697,20 +706,30 @@ func (t *tableState) findForTimestamp(
 //
 // t.mu must be locked.
 func (t *tableState) acquireFreshestFromStoreLocked(ctx context.Context, m *LeaseManager) error {
-	// Ensure there is no lease acquisition in progress.
-	t.acquireWait()
-
-	// Move forward to acquire a fresh table lease.
+	// Continue to attempt to call to acquireNodeLease until our call was the one
+	// made by singleflight. This busy-waits while another acquisitions are
+	// underway.
+	var err error
+	hasAcquiredFreshest := false
+	for !hasAcquiredFreshest {
+		// Move forward the expiry to acquire a fresh table lease.
+
+		// Set the min expiration time to guarantee that the lease acquired is the
+		// last lease in t.mu.active .
+		minExpirationTime := hlc.Timestamp{}
+		newestTable := t.mu.active.findNewest()
+		if newestTable != nil {
+			minExpirationTime = newestTable.expiration.Add(int64(time.Millisecond), 0)
+		}
 
-	// Set the min expiration time to guarantee that the lease acquired is the
-	// last lease in t.mu.active .
-	minExpirationTime := hlc.Timestamp{}
-	newestTable := t.mu.active.findNewest()
-	if newestTable != nil {
-		minExpirationTime = newestTable.expiration.Add(int64(time.Millisecond), 0)
+		t.mu.Unlock()
+		_, _, err = t.mu.group.Do("acquire", func() (interface{}, error) {
+			hasAcquiredFreshest = true
```

Comment:
I think we can do without this hasAcquiredFreshest boolean. Let's just used the shared boolean returned from Do(). It's not quite the same semantics but when set to false we are guaranteed we were the only caller of the method which means that the lease is the freshest.